### ISSUE TITLE
Provide simulation endpoints for local persistent/ephemeral volumes

### ISSUE THREAD

**Issue by ahg-g** [kubernetes/kubernetes#89312]
**What would you like to be added**:

The scheduler's [SharedLister](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/listers/listers.go#L49) interface defines APIs to retrieve listers for nodes and pods. The backing implementation used in the scheduler is based on the nodeinfo Snapshot. This scheduler.SharedLister interface is used by all plugins to retrieve and list nodes/pods, the rest of resources are accessed by the framework plugins through the general SharedInformerFactory (like NodeCSI, PVs and PVCs).

Cluster Autoscaler has logic that simulates the scheduling of pods, this logic instantiates a scheduler Framework. The Framework that CA instantiates is configured with a scheduler.SharedLister backed by CA's custom implementation that allows adding "fake" nodes to simulate possible upscale/downscale events. However, since PVs, PVCs and whatnot are listed from SharedInformerFactory, CA can't simulate those resources (e.g., simulate creating a node with local SSD etc.)

What we need is to make storage-related listers available through the scheduler framework's SharedLister interface, in the scheduler those listers will simply be backed by the SharedInformerLister, while CA can implement their own that injects fake resources.

**Why is this needed**:

To enable cluster autoscale to simulate storage-related resources.

/sig scheduling
/assign

**Comment by Huang-Wei** [comment]
SG.

A side note: the [PodInformer](https://github.com/kubernetes/kubernetes/blob/aa07992a44186452de261ba0bdd3a4afb5dfe678/cmd/kube-scheduler/app/options/options.go#L278) which internal scheduling queue relies on is not originated from SharedInformerFactory either.

Maybe we could refactor that as well so that all internal informers are unified which should benefit both default scheduler and plugin developers. WDYT?

**Comment by angao** [comment]
Any good way to get CRD resources in the scheduler framework?

**Comment by alculquicondor** [comment]
@angao your question seems off-topic for this thread. Please open a new issue if necessary.

**Comment by alculquicondor** [comment]
/remove-lifecycle stale

This is blocked on the evolution of kubernetes/enhancements#1472

cc @pohly

**Comment by pohly** [comment]
> This is blocked on the evolution of kubernetes/enhancements#1472

I wouldn't say it's blocked, but related. An initial implementation of this issue here could proceed as described and ignore storage capacity tracking.

The original description says:
> However, since PVs, PVCs and whatnot are listed from SharedInformerFactory, CA can't simulate those resources (e.g., simulate creating a node with local SSD etc.)

Why would PVCs need to be simulated?

Simulating a fake PV may be useful if that is how local SSDs are currently exposed. https://github.com/kubernetes-sigs/sig-storage-local-static-provisioner supports that, I am just not sure how commonly used it is in practice.

When using a CSI driver like PMEM-CSI or TopoLVM, local volumes will get provisioned dynamically. That then needs to simulate storage capacity for simulated nodes before the CSI driver runs on them. My idea for that was to create real `CSIStorageCapacity` objects that apply to simulated nodes, and only those. In that case, it would not be necessary to add `CSIStorageCapacity` to `SharedLister`, but we'll need certain extra labels on the simulated nodes that can be used to select the right `CSIStorageCapacity` objects.

**Comment by alculquicondor** [comment]
> I wouldn't say it's blocked, but related. An initial implementation of this issue here could proceed as described and ignore storage capacity tracking.

There is not much point on adding PV/PVC simulation capabilities to scheduler/cluster-autoscaler if it's not going to be useful for capacity tracking.

However, I agree in that simulating PVCs might not be useful at all, as opposed to PVs.

In any case, we need to collaborate in the next version of the kep.

Rephrasing title to remove implementation details

/retitle Provide simulation endpoints for local persistent/ephemeral volumes

**Comment by msau42** [comment]
cc @kubernetes/sig-storage-feature-requests

**Comment by Huang-Wei** [comment]
/reopen

I think it's still valid. Another usecase is VolumeBinding: with the PreFitlerResult's introduction, it's doable to have a fast path to check pod->pvc-pv and come up with a nodeNames slice.
